binderhub/tests/test_utils.py

```diff
@@ -0,0 +1,35 @@
+from binderhub import utils
+
+
+def test_rendezvous_rank():
+    # check that a key doesn't move if its assigned bucket remains but the
+    # other buckets are removed
+    key = "k1"
+    first_round = utils.rendezvous_rank(["b1", "b2", "b3"], key)
+    second_round = utils.rendezvous_rank([first_round[0], first_round[1]], key)
+
+    assert first_round[0] == second_round[0], key
+
+    key = "sdsdggdddddd"
+    first_round = utils.rendezvous_rank(["b1", "b2", "b3"], key)
+    second_round = utils.rendezvous_rank([first_round[0], first_round[1]], key)
+
+    assert first_round[0] == second_round[0], key
+
+    key = "crazy frog is a crazy key"
+    first_round = utils.rendezvous_rank(["b1", "b2", "b3"], key)
+    second_round = utils.rendezvous_rank([first_round[0], first_round[1]], key)
+
+    assert first_round[0] == second_round[0], key
+
+
+def test_rendezvous_independence():
+    # check that the relative ranking of two buckets doesn't depend on the
+    # presence of a third bucket
+    key = "k1"
+    two_buckets = utils.rendezvous_rank(["b1", "b3"], key)
+    three_buckets = utils.rendezvous_rank(["b1", "b2", "b3"], key)
+    # remove "b2" and check ranking
+    three_buckets.remove("b2")
+
+    assert two_buckets == three_buckets
```

Comment:
### Second test
It does the exact same thing as the first test, so it really just adds on test repetition. It does check the full ordering of the items in the list as compared to the first that only checks the first element, but checking the full ordering is very similar to checking the first element only if the list is only two in length.

- [ ] Make the lists contain some more entries to not only test that the list are the same in size as before.

Reply:
```suggestion
    assert two_buckets == three_buckets
    
def test_rendezvous_rank_perfect_stability():
    # Verify perfect stability of the algorithm experimentally. This means,
    # for a given key, to have a consistent ranking order internally among
    # a common set of buckets.

    key = "repo1"
    common_buckets = ["common_bucket_" + str(i) for i in range(80)]
    extra_buckets = ["extra_bucket_" + str(i) for i in range(20)]

    common_buckets_ranked = utils.rendezvous_rank(common_buckets, key)

    common_and_extra_buckets_ranked = utils.rendezvous_rank(common_buckets + extra_buckets, key)
    for extra_bucket in extra_buckets:
        common_and_buckets_ranked.remove(extra_bucket)

    assert common_buckets_ranked == common_and_buckets_ranked, "Adding or removing a bucket reordered the remaining buckets!"
```

Reply:
I agree that this tests essentially the same as the previous test but I thought you wanted to have a test that compares the total ranking not just the top choice which is why I added this.

Reply:
In the case with three buckets and removing one, you end up with bucket rankings with two elements each, picking out the first element determines the second anyhow, so it is pretty much the same test as the first except you verify that dimensionality isn't wierd or that you ended up with duplication of elements in the array etc.

Reply:
Aha! Thanks for the pointer to the comment on the comment. Now I think I understand better what you mean. I added a test with a large number of buckets like you suggested.

In retrospect that comment-on-the-comment was because I was confusing the case of removing and adding buckets. Since then I had become less confused and forgotten about it. I now think the comment-on-comment only applies to the adding buckets case.

I added the test you suggested as well as one for the case of adding a bucket. I'll mark this as resolved and we can start a new thread on the newest code.